pkg/specification/macspec/specification.go

```diff
@@ -0,0 +1,242 @@
+}
+
+// AllowDuplicateLinkADRAns reports whether v is allowed to use
+// duplicate LinkADRAns MAC responses within the same message.
+func AllowDuplicateLinkADRAns(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_2) >= 0 && compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) < 0
+}
+
+// AtomicLinkADRAns reports whether v will accept or reject all
+// channel masks in a singular LinkADRAns MAC response.
+func AtomicLinkADRAns(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) >= 0
+}
+
+// LimitConfirmedTransmissions reports whether v must limit
+// the number of confirmed uplink transmissions.
+func LimitConfirmedTransmissions(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_4) < 0
+}
+
+// RequireDevEUIForABP reports whether v requires ABP devices to have a DevEUI associated.
+func RequireDevEUIForABP(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_4) >= 0 && compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) < 0
+}
+
+// FrameType captures the frame type.
+type FrameType uint8
+
+const (
+	// UplinkFrame represents the uplink frame.
+	UplinkFrame FrameType = iota
+	// DownlinkFrame represents the downlink frame.
+	DownlinkFrame
+)
+
+var (
+	fOptsBlockIdentifier1 = [4]byte{0x00, 0x00, 0x00, 0x01}
+	fOptsBlockIdentifier2 = [4]byte{0x00, 0x00, 0x00, 0x02}
+)
+
+// EncryptionOptions reports the encryption options for v.
+func EncryptionOptions(v ttnpb.MACVersion, frameType FrameType, fPort uint32, cmdsInFOpts bool) (opts []crypto.EncryptionOption) {
+	// Only FOpts stored as part of the frame header require custom encryption behavior.
+	if !cmdsInFOpts {
+		return nil
+	}
+	switch cmp := compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1); {
+	case cmp < 0:
+		// FOpts are not encrypted before LoRaWAN 1.1.
+		return nil
+	case cmp >= 0:
+		// Do note that the encryption scheme described below is based on the
+		// `FOpts Encryption, Usage of FCntDwn Errata on the LoRaWAN L2 1.1 Specification`
+		// erratum, and not the literal description found in the LoRaWAN Specification 1.1.
+		var (
+			isFCntUp    = frameType == UplinkFrame
+			isNFCntDown = frameType == DownlinkFrame && fPort == 0
+			isAFCntDown = frameType == DownlinkFrame && fPort != 0
+		)
+		switch {
+		case isFCntUp, isNFCntDown:
+			return []crypto.EncryptionOption{
+				crypto.WithFrameIdentifier(fOptsBlockIdentifier1),
+			}
+		case isAFCntDown:
+			return []crypto.EncryptionOption{
+				crypto.WithFrameIdentifier(fOptsBlockIdentifier2),
+			}
+		default:
+			panic("unreachable")
+		}
+	default:
+		panic("unreachable")
+	}
+}
+
+// ValidateUplinkPayloadSize reports whether v requires that the uplink
+// payload size must be valid with respect to the regional parameters
+// and local limitations.
+func ValidateUplinkPayloadSize(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) >= 0
+}
```

Comment:
Is this a spec change since 1.1?

Reply:
Full disclosure - I added this because it existed before:
https://github.com/TheThingsNetwork/lorawan-stack/blob/70975ab7e96936103f2e1ca70b53cd68e1a64898/pkg/networkserver/downlink.go#L1825-L1832

The spec itself is quite straight forward about why we do this (page 29):
![image](https://user-images.githubusercontent.com/36161392/169009712-48021fb7-14a5-4ac3-a420-e203d3ec39bb.png)

Reply:
Okay so this is for downlink. That should be more clear.